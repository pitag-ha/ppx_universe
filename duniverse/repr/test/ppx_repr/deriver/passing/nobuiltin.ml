(* When a type is annotated with the [nobuiltin] annotation, it should be
   considered as an abstract type (i.e. don't pull representations from
   [Repr]). *)

type unit = string [@@deriving repr]

(* Shadow [Stdlib.unit] *)
module Nobuiltin_t = struct
  type t = (unit[@nobuiltin]) [@@deriving repr]

  (* [t]'s repr should be for strings. *)
  let (_ : string Repr.t) = t
end

module Nobuiltin_foo = struct
  type foo = (unit[@repr.nobuiltin]) [@@deriving repr]

  (* [foo]'s repr should be for strings too. *)
  let (_ : string Repr.t) = foo_t
end

module Nobuiltin_operator = struct
  (* Define our own representation of [result]. *)
  let result_t a b = Repr.pair a b
  let int32_t = Repr.int
  let int64_t = Repr.bool

  type u = (((int32[@nobuiltin]), int64) result[@nobuiltin]) [@@deriving repr]

  let (_ : (int * int64) Repr.t) = u_t
end
