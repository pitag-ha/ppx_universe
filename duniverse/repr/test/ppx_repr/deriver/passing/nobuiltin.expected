type unit = string[@@deriving repr]
include struct let unit_t = Repr.string end[@@ocaml.doc "@inline"][@@merlin.hide
                                                                    ]
module Nobuiltin_t =
  struct
    type t = ((unit)[@nobuiltin ])[@@deriving repr]
    include struct let t = unit_t end[@@ocaml.doc "@inline"][@@merlin.hide ]
    let (_ : string Repr.t) = t
  end
module Nobuiltin_foo =
  struct
    type foo = ((unit)[@repr.nobuiltin ])[@@deriving repr]
    include struct let foo_t = unit_t end[@@ocaml.doc "@inline"][@@merlin.hide
                                                                  ]
    let (_ : string Repr.t) = foo_t
  end
module Nobuiltin_operator =
  struct
    let result_t a b = Repr.pair a b
    let int32_t = Repr.int
    let int64_t = Repr.bool
    type u = (((((int32)[@nobuiltin ]), int64) result)[@nobuiltin ])[@@deriving
                                                                    repr]
    include struct let u_t = result_t int32_t Repr.int64 end[@@ocaml.doc
                                                              "@inline"]
    [@@merlin.hide ]
    let (_ : (int * int64) Repr.t) = u_t
  end
